### ISSUE TITLE
Gcp Provider does not use gcp api from correct project

### ISSUE THREAD

**Issue by gerethd** [hashicorp/terraform-provider-google#13298]
Running from jenkins where the default project is set to project A via gcloud cli

resources where the provider is set to project B still use the api quotas from project A resulting in errors that shouldn't happen when running larger projects.

**Comment by edwardmedia** [comment]
@gerethd I don't know exactly how you organize your code. Have you tried to provide project to individual resource (instead of set it at the provider level?

**Comment by gerethd** [comment]
It is set at the resource level, this is specifically an issue for database
instances with many small databases in them

Regards,
Gereth Dittrick

**Comment by edwardmedia** [comment]
@gerethd can you share the config that can used to repro the issue along with the debug log?

**Comment by gerethd** [comment]
I will put together a test repo for this, unfortunately it's corporate code
so i can't just link the existing config

Regards,
Gereth Dittrick

**Comment by edwardmedia** [comment]
@gerethd waiting for code to repro

**Comment by gerethd** [comment]
https://github.com/zevrant/terraform-gcp-test

it's important that the tf state be stored in a project C while most resources are in project A and databases in project B and all be accessible from the same deploy account

**Comment by gerethd** [comment]
the issue is that terraform uses the apis from project C instead of A and B where the resources are created

**Comment by edwardmedia** [comment]
@gerethd I opened your repo, and it uses other modules which is invisible to me. I can't apply it to repro the issue.

Please keep in mind this repo owns the provider which handles individual resource and data source. I am not sure if this is the right place that can help address your question. It sounds to me a question it is the terraform core/sdk that can explain better.

Can you share your debug log that can explain what the issue you are facing?

**Comment by gerethd** [comment]
this is specifically the gcp terraform provider as mentioned before, the provider is running all api queries using project C instead of A and B where the resources are created, i'm not sure how that would be a core terraform issue.

Idk if i can provide a log dump without exposing corporate information, i will generate one and purge it of information specific to our resources

**Comment by edwardmedia** [comment]
@gerethd this seems a very common use case. It doesn't matter where you run the `terraform apply`. It could be in a gcp project, like you said, project C or a local machine. What matters is the [account](https://registry.terraform.io/providers/hashicorp/google/latest/docs/guides/provider_reference#authentication) that you use to run `terraform apply` and that account needs to have certain permissions to project A and B. 

I do not understand what you said below. You can't create a resource in Project A and B without enabling the APIs at the Project A and B. 

```
the provider is running all api queries using project C instead of A and B where the resources are created
```

Possibly I could misunderstood the issue. That is why I want to see your config and want to repro the issue.

Does this make sense?

**Comment by gerethd** [comment]
I cannot give you the config as requested, company security says no.

The account that is used to deploy essentially has admin access in all three accounts.

I'm running all of this from a single location in parallel so location is not the question.

The api endpoints that the provider uses to query the current state of gcp resources is the api queries i'm referring to. Each of our projects has a separate quota.

Project A has a quota of 180
Project B has a quota of 1000 (database project)
Project C has a quota of 180

Resources are queried and created in projects A and B but use the quotas from Project C to do so

**Comment by edwardmedia** [comment]
@gerethd I do not understand the issue. I don't know what relationships between Project A and B and C. If they are shared VPCs, (vs independent projects)  you may want to review the related docs. Certain rules are applied.

My assumption is you run `terraform apply` in project C. Can you try to run it at the Project A instead, even local to see the issue persists?

**Comment by gerethd** [comment]
i'm not running terraform apply from within any of the projects

**Comment by gerethd** [comment]
Projects A, B, and C are all on the same network and is inside our corporate firewall. I myself can't actually see any more than that.

That doesn't change the point that if i set the provider to use Project B that it should use the APIs/Quota from project B

**Comment by edwardmedia** [comment]
@megan07 what do you think about this issue?

**Comment by megan07** [comment]
Hi @gerethd, I'm sorry you're running into this issue and for the misunderstanding. Of course, raw configuration and debug logs would be helpful, but I completely understand you're not allowed to share them.

I'm a little confused where all the projects are coming into play. Would you be able to verify something simple like this?

```
provider "google" {
	project = "project_c"
}

resource "google_database_instance" "foo" {
	project = "project_b"
}

resource "google_storage_bucket" "bar" {
	project = "project_a"
}
```

Do you have any environment variables set (such as `GOOGLE_PROJECT`) that I should be aware of? Also, can you clarify which project you're expecting the quotas from for each resource as well as where they're _actually_ coming from (for each resource).

Thanks!

**Comment by gerethd** [comment]
This is being run in a docker container with the only environment variable set being GOOGLE_OAUTH_ACCESS_TOKEN, and there is no pre configuration done at all, i.e. fresh install of gcloud with none of the initialization being done, prior to runtime. Unless the federated access token to access GCP itself has a default project all configs should be blank or whatever the default value is.

project_c is never explicitly referenced in the code, that's just where the cloud storage bucket is located. All of our environments store the terraform state in this bucket. In fact the project id does not exist in the project itself like in the example repo i included. I'm guessing by providing the name of the bucket, and that bucket only being found in project C defaulting to that project.

the full structure looks like this, and we use a single deploy account to deploy to all environments


project C -> common project to hold resources needed to run terraform

project A -> preprod project holding preprod environments (dev, test etc.)
project B -> preprod project used for holding data storage resources for preprod environments  (dev, test etc.)

project D - prod project holding prod like environments (stage, prod)
project E - prod project holding data storage resources for prod like environments (stage, prod)

when running for an environment such as dev we have the provider set to project A, and the terraform state in project C, which as mentioned above that project id appears to be autoresolved based on the buckets the deploy account has access to. The dev environment has resources created in both projects A and B, where resources in B we use the project label on the resource itself like [this](https://github.com/zevrant/terraform-gcp-test/blob/7c6408213d05503caf9b03a6afce52d8be506ce8/modules/templates/db-instance/main.tf#L7) to deploy it into project B while still having the provider set to project A. 

The ideal behavior would be that resources deployed into project B via that flag would use the quotas for project B, the resources deployed to project A by having the provider set to project A such as [this here](https://github.com/zevrant/terraform-gcp-test/blob/7c6408213d05503caf9b03a6afce52d8be506ce8/modules/environment/providers.tf#L2) would use the quotas from project A.

At minimum having quotas used from the provider config in the parent module that is set to project A [here](https://github.com/zevrant/terraform-gcp-test/blob/7c6408213d05503caf9b03a6afce52d8be506ce8/modules/environment/providers.tf#L2) would cause sub modules to use those quotas instead of Project C, since the only purpose for project C is to hold terraform state and related materials

**Comment by megan07** [comment]
Ah, ok, so we're looking more at something like this (configuration wise):
```
provider "google" {
	project = "project_a"
}

resource "google_database_instance" "foo" {
	project = "project_b"
}

resource "google_<any_other_resource>" "bar" {
	<inherits project from provider>
}
```

and then the state is stored in a bucket in project_c, is that correct?

and you've confirmed that all the resources are using project_c instead of the ones you've assigned in the config? can you confirm this in debug logs?

i'm going to try and reproduce the issue here to make sure i've got it and i can better help you, sorry for the additional clarification.

**Comment by gerethd** [comment]
yes this is correct, no worries i'd rather us all be on the same page first

**Comment by gerethd** [comment]
I don't need to confirm in the debug logs as we are exceeding the quota for our sql admin queries and all of our sql resources are tagged as the same as resource `foo` in your example and the error message says that the quota is exceeded for project C
